### DISCUSSION TITLE
[Enhancement] [mention] 希望提及Mention组件新增checkVisible属性

### DISCUSSION THREAD

**Discussion by SorrowX** [element-plus/element-plus#18990] [Feature Request]
Feature type: **`Enhancement`**

## Details

- Enhanced component name: `mention`

## Description

需求：希望Mention组件提及时可以限制提及的最大次数，真实业务是这样的，我们在使用Mention组件时，产品希望只@一个人，后面再次@时，提及弹窗不在出现
实现：
方案一：Mention组件组件新增limit属性，用来控制提及最大的次数。但是实现时，需要在其内部维护已经提及的列表(包含添加和删除操作)，根据维护好的提及列表和这个新增属性做判断就能实现该功能。
优点：内部维护了提及列表，且能抛出给我们用户层使用，编写业务代码时，代码量更加精简
缺点：代码改动较大

方案二:Mention组件组件新增checkVisible 函数属性，用于唤起弹窗时执行该钩子，让用户自行判断判断是否需要唤起。
优点：代码改动少，无入侵式代码，且更加灵活
缺点：外部需要自行维护提及的列表(增加和删除，而且删除事件还没有，只能根据check-is-whole函数自行扩展)

我选择了方案二 使用姿势如下：
```vue
<template>
  <div>
    <el-mention
      ref="mentionRef"
      whole
      v-model="html"
      :options="options"
      style="width: 320px"
      placeholder="Please input"
      :checkVisible="handleCheckVisible"
      @select="handleSelect"
    />
  </div>
</template>

<script setup lang="ts">
import { ref } from 'vue'
import type { MentionOption } from 'element-plus'

const html = ref('')
const mentionRef = ref()

const options = ref([
  {
    label: 'Fuphoenixes',
    value: 'Fuphoenixes',
  },
  {
    label: 'kooriookami',
    value: 'kooriookami',
  },
  {
    label: 'Jeremy',
    value: 'Jeremy',
  },
  {
    label: 'btea',
    value: 'btea',
  },
])

let atList: MentionOption[] = []

const handleSelect = (option: MentionOption) => {
  const item = options.value.find(
    (item: MentionOption) => item.value === option.value
  )
  if (item) {
    atList.push(item)
  }
}

// 最大允许@ 2个人
const handleCheckVisible = (
  visible: boolean,
  filteredOptions: MentionOption[],
  mentionCtx: any
) => {
  console.log('check: ', visible, filteredOptions, mentionCtx)
  return visible && atList.length < 2
}
</script>

```

希望官方可以增加这个特性

### Additional comments

_(empty)_

**Comment by btea** [comment]
@SorrowX  新加一个需求流程可能没那么快，基于你的需求，我写了一个简单的示例，可供参考。[link](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZWwtbWVudGlvblxuICAgIHdob2xlXG4gICAgdi1tb2RlbD1cInZhbHVlXCJcbiAgICA6b3B0aW9ucz1cIm9wdGlvbnNcIlxuICAgIHN0eWxlPVwid2lkdGg6IDMyMHB4XCJcbiAgICBwbGFjZWhvbGRlcj1cIlBsZWFzZSBpbnB1dFwiXG4gIC8+XG48L3RlbXBsYXRlPlxuXG48c2NyaXB0IHNldHVwIGxhbmc9XCJ0c1wiPlxuaW1wb3J0IHsgcmVmLCB3YXRjaCB9IGZyb20gJ3Z1ZSdcbmNvbnN0IHZhbHVlID0gcmVmKCdAJylcblxuY29uc3QgaW5pdE9wdGlvbnMgPSBbXG4gIHtcbiAgICBsYWJlbDogJ0Z1cGhvZW5peGVzJyxcbiAgICB2YWx1ZTogJ0Z1cGhvZW5peGVzJyxcbiAgfSxcbiAge1xuICAgIGxhYmVsOiAna29vcmlvb2thbWknLFxuICAgIHZhbHVlOiAna29vcmlvb2thbWknLFxuICB9LFxuICB7XG4gICAgbGFiZWw6ICdKZXJlbXknLFxuICAgIHZhbHVlOiAnSmVyZW15JyxcbiAgfSxcbiAge1xuICAgIGxhYmVsOiAnYnRlYScsXG4gICAgdmFsdWU6ICdidGVhJyxcbiAgfSxcbl1cbmNvbnN0IG9wdGlvbnMgPSByZWYoaW5pdE9wdGlvbnMpXG5cbndhdGNoKCgpID0+IHZhbHVlLnZhbHVlLCAodikgPT4ge1xuICBpZiAodikge1xuICAgIG9wdGlvbnMudmFsdWUgPSBbXVxuICB9IGVsc2Uge1xuICAgIG9wdGlvbnMudmFsdWUgPSBpbml0T3B0aW9uc1xuICB9XG59KVxuPC9zY3JpcHQ+XG4iLCJlbGVtZW50LXBsdXMuanMiOiJpbXBvcnQgRWxlbWVudFBsdXMgZnJvbSAnZWxlbWVudC1wbHVzJ1xuaW1wb3J0IHsgZ2V0Q3VycmVudEluc3RhbmNlIH0gZnJvbSAndnVlJ1xuXG5sZXQgaW5zdGFsbGVkID0gZmFsc2VcbmF3YWl0IGxvYWRTdHlsZSgpXG5cbmV4cG9ydCBmdW5jdGlvbiBzZXR1cEVsZW1lbnRQbHVzKCkge1xuICBpZiAoaW5zdGFsbGVkKSByZXR1cm5cbiAgY29uc3QgaW5zdGFuY2UgPSBnZXRDdXJyZW50SW5zdGFuY2UoKVxuICBpbnN0YW5jZS5hcHBDb250ZXh0LmFwcC51c2UoRWxlbWVudFBsdXMpXG4gIGluc3RhbGxlZCA9IHRydWVcbn1cblxuZXhwb3J0IGZ1bmN0aW9uIGxvYWRTdHlsZSgpIHtcbiAgY29uc3Qgc3R5bGVzID0gWydodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL2VsZW1lbnQtcGx1c0BsYXRlc3QvZGlzdC9pbmRleC5jc3MnLCAnaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9lbGVtZW50LXBsdXNAbGF0ZXN0L3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Nkbi5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9kaXN0L2luZGV4LmZ1bGwubWluLm1qc1wiLFxuICAgIFwiZWxlbWVudC1wbHVzL1wiOiBcImh0dHBzOi8vY2RuLmpzZGVsaXZyLm5ldC9ucG0vZWxlbWVudC1wbHVzQGxhdGVzdC9cIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9jZG4uanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6e319)

**Reply by SorrowX** [reply]
需求流程不快也没关系，只求后面能加上即可

**Reply by SorrowX** [reply]
感谢大佬提供的思路，我可以根据收集的提及信息来把options选项是否重置为空，确实可以，谢谢大佬
不过我提的这个诉求，应该也没啥毛病

**Reply by btea** [reply]
是的，我只是基于目前的情况提供一个解决问题的思考方向。组件内部能通过配置实现当然对用户更友好。

**Reply by SorrowX** [reply]
大佬，还有诉求，就是当使用whole属性可以整体删除时，希望抛出个delete事件，该事件方便我们用户把本地收集的提及列表根据delete事件再删除

**Reply by btea** [reply]
看到了相关的pr，还需要其他相关人员的审核，进度可能没那么快，请耐心等待。

**Reply by SorrowX** [reply]
> 看到了相关的pr，还需要其他相关人员的审核，进度可能没那么快，请耐心等待。

没事，我可以等，只要后期可以加上该api即可，如果需要配合啥的，我全力配合

**Comment by SorrowX** [comment]
![image](https://github.com/user-attachments/assets/18d174ac-ab51-4e0d-8e3b-36002b20747c)
我目前是根据这2个属性来维护收集提及的列表，这么做会有一个问题，首先我们@的选项和选中的选项希望是一样的，给后端时是id,所以我这么做，当出现同名时，删除的就是第一个，也是个bug,所以希望后面添加一个delete事件，来解决我目前的问题

**Reply by btea** [reply]
有点没看懂你这里描述的问题和需求，能更加详细一点吗？

**Reply by SorrowX** [reply]
稍等，我来详细描述下

**Reply by SorrowX** [reply]
![image](https://github.com/user-attachments/assets/b804ccd3-021d-43db-b349-4a6d0d1d94f0)
就拿这个图里面的提及组件，可以看的出来提及列表，展示的和选中后的都是相同的名字，对应我们的提及组件options只能 
[{ label: ‘兮兮酱’, value: '兮兮酱' }, ...] ，但是给后端的时候的肯定不能给label，需要给字典值。
所以这时我需要把options扩展成 [{ label: ‘兮兮酱’, value: '兮兮酱',id: '1111' }, ...]这种形式，我根据select事件收集option时，比较简单，直接把option push到数组中，但是删除时，目前根据checkIsWhole 属性回调，而且参数只有2个可用
![image](https://github.com/user-attachments/assets/348ae4d5-e1c9-4958-8302-e12908fcc5b1)
根据源代码，我记得pattern是value值，然而为了选中和回显是同一个值，我把label和value都写成一样了，这时候根据上面这图，如果出现同名value时，删除则会出现问题。
解决方案：
方案1： checkIsWhole函数参数再抛出一个matchOption
```js
 const matchOption = props.options.find((item) => item.value === pattern)
        const isWhole = isFunction(props.checkIsWhole)
          ? props.checkIsWhole(pattern, prefix, matchOption)
          : matchOption
```
这样也能解决问题，但是还得把getMentionCtx函数通过expose导出
方案2： 
抛出一个delete事件，最简单

**Reply by btea** [reply]
看起来是有些复杂。

**Reply by SorrowX** [reply]
可能我表达的不够清楚😂
